loom/graphics/gfxShader.cpp

```diff
@@ -0,0 +1,509 @@
+
+    programId = ctx->glCreateProgram();
+
+    // Store local pointers and lengths because of the GL API
+    const GLchar *vssp = static_cast<const GLchar*>(vss);
+    const GLint vssl = strlen(vss);
+    const GLchar *fssp = static_cast<const GLchar*>(fss);
+    const GLint fssl = strlen(fss);
+
+    // Compile VS and FS
+    ctx->glShaderSource(vertexShaderId, 1, &vssp, &vssl);
+    Graphics::context()->glCompileShader(vertexShaderId);
+    GFX_SHADER_CHECK(vertexShaderId);
+
+    ctx->glShaderSource(fragmentShaderId, 1, &fssp, &fssl);
+    ctx->glCompileShader(fragmentShaderId);
+    GFX_SHADER_CHECK(fragmentShaderId);
+
+    // Link the program
+    ctx->glAttachShader(programId, fragmentShaderId);
+    ctx->glAttachShader(programId, vertexShaderId);
+    ctx->glLinkProgram(programId);
+    GFX_PROGRAM_CHECK(programId);
+
+    // Lookup vertex attribute array locations
+    posAttribLoc = Graphics::context()->glGetAttribLocation(programId, "a_position");
+    posColorLoc = Graphics::context()->glGetAttribLocation(programId, "a_color0");
+    posTexCoordLoc = Graphics::context()->glGetAttribLocation(programId, "a_texcoord0");
+}
+
+void GFX::Shader::loadFromAssets(const char* vertexShaderPath, const char* fragmentShaderPath)
+{
+    void * vertData = loom_asset_lock(vertexShaderPath, LATVertexShader, 1);
+    if (vertData == NULL)
+    {
+        lmLogWarn(gGFXShaderLogGroup, "Unable to lock the asset for shader %s", vertexShaderPath);
+        return;
+    }
+    loom_asset_unlock(vertexShaderPath);
+
+    void * fragData = loom_asset_lock(fragmentShaderPath, LATFragmentShader, 1);
+    if (fragData == NULL)
+    {
+        lmLogWarn(gGFXShaderLogGroup, "Unable to lock the asset for shader %s", vertexShaderPath);
+        return;
+    }
+    loom_asset_unlock(vertexShaderPath);
+
+    const char *vert = static_cast<char *>(vertData);
+    const char *frag = static_cast<char *>(fragData);
+
+    load(vert, frag);
+}
+
+GLint GFX::Shader::getUniformLocation(const char* name)
+{
+    GFX::GL_Context* ctx = Graphics::context();
+    return ctx->glGetUniformLocation(programId, name);
+}
+
+#include "loom/common/platform/platformTime.h"
+
+void GFX::Shader::setUniform1f(GLint location, GLfloat v0)
+{
+    Graphics::context()->glUniform1f(location, v0);
+}
+
+int GFX::Shader::setUniform1fv(lua_State *L)
+{
+    GLint location = (GLint)lua_tonumber(L, 2);
+    int length = lsr_vector_get_length(L, 3);
+
+    utArray<float> values;
+
+    lua_rawgeti(L, 3, LSINDEXVECTOR);
+    int vidx = lua_gettop(L);
+
+    for (int i = 0; i < length; i++)
+    {
+        lua_rawgeti(L, vidx, i);
+        float value = (float)lua_tonumber(L, -1);
+        values.push_back(value);
+    }
+
+    Graphics::context()->glUniform1fv(location, values.size(), values.ptr());
+    return 0;
+}
+
+void GFX::Shader::setUniform2f(GLint location, GLfloat v0, GLfloat v1)
+{
+    Graphics::context()->glUniform2f(location, v0, v1);
```

Comment:
How would we do sanity testing?

I think that buffering would be an unnecessary (potential) performance hit